### PR TITLE
feat: Application Insights Extensions with many features

### PR DESCRIPTION
**Pull request by darthkurak**
Please, read Readme.md first, as it should ease understanding whole PR

### CODE REVIEW

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/nuget.config
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/nuget.config:1]: Not a typical place for a nuget config (in a project folder), it's more often put in a solution folder, any reason this is here?

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/Allegro.Extensions.ApplicationInsights.AspNetCore.csproj [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        <TargetFramework>net6.0</TargetFramework>
+        <Nullable>enable</Nullable>
+        <IsPackable>true</IsPackable>
+    </PropertyGroup>
+    <Import Project="..\..\Package.Build.props" />
+    <ItemGroup>
+      <PackageReference Include="FluentAssertions" Version="6.7.0" />
+      <PackageReference Include="Microsoft.ApplicationInsights" Version="2.21.0" />
+      <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.21.0" />
+      <PackageReference Include="Microsoft.AspNetCore.Hosting.Abstractions" Version="2.2.0" />
```
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/Allegro.Extensions.ApplicationInsights.AspNetCore.csproj:13]: This is an old package, the `Microsoft.Extensions.Hosting.Abstractions` replaced it - why the reference?

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Demo/Controllers/WeatherForecastController.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+using System.Net.Http;
+using Allegro.Extensions.ApplicationInsights.AspNetCore;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
```
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Demo/Controllers/WeatherForecastController.cs:6]: Maybe just disable for the whole project, it makes sense for Demo not to be doc'ed. `<NoWarn>$(NoWarn);CS1591</NoWarn>

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Demo/InternalsVisibleTo.cs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo(
+    "Allegro.Extensions.ApplicationInsights.Demo.UnitTests")]
```
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Demo/InternalsVisibleTo.cs:4]: Can be also done in project file in ItemGroup `<InternalsVisibleTo Include="..." />` - less `.cs` files!

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Prometheus.UnitTests/Usings.cs [hunk 4]
```diff
@@ -0,0 +1 @@
+global using Xunit;
```
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Prometheus.UnitTests/Usings.cs:1]: Can also be done in project file (less codeless .cs files): `<Using Include="Xunit" />` in ItemGroup.

File: src/Allegro.Extensions.ApplicationInsights/README.md [hunk 5]
```diff
@@ -0,0 +1,468 @@
+# Allegro.Extensions.ApplicationInsights
+
+This repo contains a collection of useful Azure Application Insights extensions and utilities, developed as part of [Allegro Pay](https://allegropay.pl/) product which allow you:
```
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/README.md:3]: > product which allow you:

Last part doesn't make sense. I'd suggest just `product.`. The following is a ToC, and it doesn't read well:

> extensions [...] which allow you: - better sampling configuration.

File: src/Allegro.Extensions.ApplicationInsights/README.md [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+
+Just add:
+
+```csharp
+builder.AddApplicationInsightsSamplingConfig()
+```
+
+and now you can configure using such sections (represent default values):
+
+```json
```
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/README.md:37]: Might make the comment to render without the red/error background:
```suggestion
```jsonc
```

File: src/Allegro.Extensions.ApplicationInsights/README.md [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+The library has some protection against it which you can enable and get a proper `operation_name` and target url (`name`)
+
+Firstly, it has a built-in deparametrizator that tries to find query parameter values and change them to constants. So each query path will be recorded as the same label value, despite different param values.
+You can enable it with `ShouldGeneralizeHttpDependencyOperationName` or `ShouldGeneralizeHttpDependencyTargetUrl`
+
+Secondly, if it fails and finds more than `MaxUrisPerHost` (100 by default) - it will circuit break for that host and log an error to `IStartupLogger` (you need to enable it)
+
+However, despite these protections, <span style="color:red;font-size:20px">BE CAUCIUS</span>, when using this feature. Check your metrics cardinality and size.
+
+## Bringing all together
```
Comment (amis92) [inline src/Allegro.Extensions.ApplicationInsights/README.md:413]:
```suggestion
## Bringing it all together
```

**Comment by ppiwow-apay** [comment]
add entry on main readme about new util.

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    public static void CompilingRulesFailed(this ILogger logger, Exception exception, string filterKey, string filterValue, string filterType)
+    {
+        _compilingRulesFailed(logger, filterKey, filterValue, filterType, exception);
+    }
+}
+
+/// <summary>
+/// It is publicly visible for unit tests purposes. See Demo.UnitTests
```
Comment (ppiwow-apay) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs:36]: internals visible to in tests?

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    private readonly Func<DependencyTelemetry, TDependencyForFilter> _dependencyMap;
+    private readonly List<Func<TDependencyForFilter, bool>> _dependencyPredicates;
+    private readonly ILogger _logger;
+    private readonly Func<RequestTelemetry, TRequestForFilter> _requestMap;
+    private readonly List<Func<TRequestForFilter, bool>> _requestPredicates;
+
+    /// <summary>
+    /// It is publicly visible for unit tests purposes. See Demo.UnitTests
+    /// </summary>
+    public ExcludeFromSamplingTelemetryInitializer(
```
Comment (ppiwow-apay) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs:51]: internals visible to

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Prometheus/InternalsVisibleTo.cs [hunk 10]
```diff
@@ -0,0 +1,3 @@
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Allegro.Extensions.ApplicationInsights.Prometheus.UnitTests")]
```
Comment (ppiwow-apay) [inline src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Prometheus/InternalsVisibleTo.cs:3]: put in csproj?

**Review by ppiwow-apay — APPROVED** [review]

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Demo/InternalsVisibleTo.cs (hunk 3 above)
Comment (darthkurak) [reply src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Demo/InternalsVisibleTo.cs:4]: I prefer this way. I used to it, and it is more visible and convinient to change

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs (hunk 8 above)
Comment (darthkurak) [reply src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs:36]: Nope, it is for people, to write their own tests.

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Prometheus/InternalsVisibleTo.cs (hunk 10 above)
Comment (darthkurak) [reply src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.Prometheus/InternalsVisibleTo.cs:3]: i prefer via file :)

File: src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs (hunk 9 above)
Comment (darthkurak) [reply src/Allegro.Extensions.ApplicationInsights/Allegro.Extensions.ApplicationInsights.AspNetCore/SamplingExclusions/ExcludeFromSamplingTelemetryInitializer.cs:51]: Nope, it is for people, to write their own tests.

**Pull request merged**